app/actions/local/thread.ts

```diff
@@ -62,10 +95,236 @@ export const switchToThread = async (serverUrl: string, rootId: string) => {
+    }
+
+    return {models};
+};
+
+// On receiving threads, Along with the "threads" & "thread participants", extract and save "posts" & "users"
+export const processReceivedThreads = async (serverUrl: string, threads: Thread[], prepareRecordsOnly = false) => {
+    const operator = DatabaseManager.serverDatabases[serverUrl]?.operator;
+    if (!operator) {
+        return {error: `${serverUrl} database not found`};
+    }
+
+    const models: Model[] = [];
+
+    const posts: Post[] = [];
+    const users: UserProfile[] = [];
+
+    // Extract posts & users from the received threads
+    for (let i = 0; i < threads.length; i++) {
+        const {participants, post} = threads[i];
+        posts.push(post);
+        participants.forEach((participant) => users.push(participant));
+    }
+
+    const postModels = await operator.handlePosts({
+        actionType: ActionType.POSTS.RECEIVED_IN_CHANNEL,
+        order: [],
+        posts,
+        prepareRecordsOnly: true,
+    });
+
+    if (postModels.length) {
+        models.push(...postModels);
+    }
+
+    const threadModels = await operator.handleThreads({
+        threads,
+        prepareRecordsOnly: true,
+    });
+
+    if (threadModels.length) {
+        models.push(...threadModels);
+    }
+
+    const userModels = await operator.handleUsers({
+        users,
+        prepareRecordsOnly: true,
+    });
+
+    if (userModels.length) {
+        models.push(...userModels);
+    }
+
+    if (models.length && !prepareRecordsOnly) {
+        await operator.batchRecords(models);
+    }
+    return {models};
+};
+
+export const processUpdateTeamThreadsAsRead = async (serverUrl: string, teamId: string, prepareRecordsOnly = false) => {
+    const operator = DatabaseManager.serverDatabases[serverUrl]?.operator;
+    if (!operator) {
+        return {error: `${serverUrl} database not found`};
+    }
+    try {
+        const {database} = operator;
+        const threads = await queryThreadsInTeam({database, teamId}).fetch();
+        const models = threads.map((thread) => thread.prepareUpdate((record) => {
+            record.unreadMentions = 0;
+            record.unreadReplies = 0;
+        }));
+        if (!prepareRecordsOnly) {
+            await operator.batchRecords(models);
+        }
+        return {models};
+    } catch (error) {
+        return {error};
+    }
+};
+
+export const processUpdateThreadFollow = async (serverUrl: string, threadId: string, state: boolean, replyCount?: number, prepareRecordsOnly = false) => {
+    const operator = DatabaseManager.serverDatabases[serverUrl]?.operator;
+    if (!operator) {
+        return {error: `${serverUrl} database not found`};
+    }
+
+    try {
+        const {database} = operator;
+        const thread = await getThreadById(database, threadId);
+        if (thread) {
+            const model = thread.prepareUpdate((record) => {
+                record.replyCount = replyCount ?? record.replyCount;
```

Comment:
When a post is deleted or created, we get `reply_count` of the root post. In case you are not following the thread and but you are in a channel and some one replies to those threads, we will capture post `delete` and `posted` events and update the `replyCount`